palm/plugins/dbt/commands/cmd_run.py

```diff
@@ -1,60 +1,155 @@
+@click.option("--iterative", is_flag=True, help="Iterative stateful dbt run")
 @click.option(
     "--full-refresh",
     is_flag=True,
-    help="will perform a full refresh on incremental models",
+    help="Will perform a full refresh on incremental models",
 )
-@click.option("--no-seed", is_flag=True, help="will skip seed full refresh")
-@click.pass_context
+@click.option("--no-seed", is_flag=True, help="Will skip seed full refresh")
+@click.pass_obj
 def cli(
-    ctx,
+    environment,
     no_fail_fast: bool,
     clean: bool,
     full_refresh: bool,
+    iterative: bool,
+    defer: bool,
     no_seed: bool,
     models: Optional[Tuple] = tuple(),
     select: Optional[Tuple] = tuple(),
     exclude: Optional[Tuple] = tuple(),
+    vars: Optional[str] = None,
 ):
     """Runs the DBT repo."""
+    stateful = iterative or defer
+
+    if defer:
+        click.secho("Running 'palm prod-artifacts'...", fg='yellow')
+        exit_code, _, _ = environment.run_on_host("palm prod-artifacts")
+        if exit_code == 2:
+            click.secho(
+                "'palm prod-artifacts' not implemented. Can't pull prod artifacts without it!",
+                fg='red',
+            )
+            sys.exit(1)
+        elif exit_code != 0:
+            click.secho(
+                "Something went wrong while pulling the prod artifacts.", fg='red'
+            )
+            sys.exit(1)
+        artifacts = 'prod'
+    else:
+        artifacts = 'local'
+
+    env_vars = set_env_vars(environment, stateful, artifacts)
+
+    # --select and --models are interchangeable on dbt >= v1, combine the lists of selections
+    targets = list(set(models + select))
+
+    # Build model run command
+    run_cmd = build_run_command(
+        # Is there a better way to pass these args in?
+        full_refresh=full_refresh,
+        no_seed=(no_seed or defer),
+        no_fail_fast=(no_fail_fast or iterative),
+        targets=targets,
+        exclude=exclude,
+        defer=defer,
+        vars=vars,
+    )
 
+    # Where the magic happens
+    success, msg = environment.run_in_docker(run_cmd, env_vars)
+
+    if iterative:
+        while not success:
+            click.secho("dbt run had some issues. Fix, then hit continue.", fg="yellow")
+            cycle = click.confirm("Continue?", show_default=True)
+
+            if not cycle:
+                break
+
+            env_vars = set_env_vars(environment, stateful, artifacts='local')
+            stateful_run_cmd = build_run_command(
+                targets=["result:error", "result:skipped"],
+                no_seed=iterative,
+                no_fail_fast=iterative,
+            )
+            success, msg = environment.run_in_docker(stateful_run_cmd, env_vars)
+
+    click.secho(msg, fg="green" if success else "red")
+
+    if clean:
+        # Clean up after ourselves
+        success, msg = environment.run_in_docker(
+            "dbt run-operation drop_branch_schemas && dbt clean", env_vars
```

Comment:
Will `env_vars` be correct here if we just finished a run using `--iterative --defer --clean`? The first run would use `artifacts='prod'`, then subsequent runs would use `artifacts='local'` so our env vars would contain `env_vars['DBT_ARTIFACT_STATE_PATH'] = plugin_config.dbt_artifacts_local`, would that leave anything hanging out after clean?

Reply:
Like mentioned in your other comment, `dbt clean` uses the `clean-targets` variable in the dbt_project.yml file, not the env vars! I don't believe many projects even use the env vars because you have to configure your environment to assign them. Most folks use the flags on the dbt commands to specify `--state path/to/artifacts`!